datumaro/plugins/open_images_format.py

```diff
@@ -349,38 +349,113 @@ def find_sources(cls, path):
+            self._writer_manager.__exit__(exc_type, exc_value, traceback)
+
+    def writerow(self, rowdict):
+        if not self._writer_manager:
+            self._writer_manager = self._writer_manager_factory()
+            self._writer = self._writer_manager.__enter__()
+
+        self._writer.writerow(rowdict)
+
+
+class _AnnotationWriter:
+    _POSSIBLE_ANNOTATION_FILE_PATTERNS = (
+        # class description files don't need to be listed,
+        # because they are always written
+        OpenImagesPath.FULL_IMAGE_DESCRIPTION_FILE_NAME,
+        *OpenImagesPath.SUBSET_IMAGE_DESCRIPTION_FILE_PATTERNS,
+        '*' + OpenImagesPath.LABEL_DESCRIPTION_FILE_SUFFIX,
+        '*' + OpenImagesPath.BBOX_DESCRIPTION_FILE_SUFFIX,
+        DEFAULT_IMAGE_META_FILE_NAME,
+    )
 
-        with open(absolute_path, 'w', encoding='utf-8', newline='') as f:
-            yield csv.DictWriter(f, field_names)
+    def __init__(self, root_dir):
+        self._annotations_dir = osp.join(root_dir, OpenImagesPath.ANNOTATIONS_DIR)
+        self._written_annotations = set()
 
-    def apply(self):
-        annotations_dir = osp.join(self._save_dir, OpenImagesPath.ANNOTATIONS_DIR)
+        os.makedirs(self._annotations_dir, exist_ok=True)
 
-        os.makedirs(annotations_dir, exist_ok=True)
+    def open(self, file_name, newline=None):
+        self._written_annotations.add(file_name)
+        f = open(
+            osp.join(self._annotations_dir, file_name),
+            'w', encoding='utf-8', newline=newline)
+        return f
 
-        self._save_categories()
-        self._save_label_category_parents()
-        self._save_subsets()
+    @contextlib.contextmanager
+    def open_csv(self, file_name, field_names, *, write_header=True):
+        with self.open(file_name, newline='') as f:
+            writer = csv.DictWriter(f, field_names)
+            if write_header:
+                writer.writeheader()
+            yield writer
+
+    def open_csv_lazy(self, file_name, field_names):
+        return _LazyCsvDictWriter(
+            lambda: self.open_csv(file_name, field_names))
+
+    def remove_unwritten(self):
+        for file_name in os.listdir(self._annotations_dir):
+            if file_name not in self._written_annotations and any(
+                fnmatch.fnmatch(file_name, pattern)
+                for pattern in self._POSSIBLE_ANNOTATION_FILE_PATTERNS
+            ):
+                os.unlink(osp.join(self._annotations_dir, file_name))
+
+class OpenImagesConverter(Converter):
+    DEFAULT_IMAGE_EXT = '.jpg'
 
-    def _save_categories(self):
-        with self._open_csv_annotation(
+    def apply(self):
+        self._save(_AnnotationWriter(self._save_dir))
+
+    @classmethod
+    def patch(cls, dataset, patch, save_dir, **options):
+        converter = cls(dataset, save_dir, **options)
+        annotation_writer = _AnnotationWriter(save_dir)
+        converter._save(annotation_writer)
+        annotation_writer.remove_unwritten()
```

Comment:
Look at other implementations. Probably, I've found a more or less generic heuristic for removal of subsets - [here](https://github.com/openvinotoolkit/datumaro/pull/365/files#diff-13eb8f857f20c388f388dd802efed6d9f85486bc3294cb55b930cf7200766544R310-R315).

Reply:
This approach only works if you can accurately predict the file name from which the deleted subsets' annotations were loaded. Which, for this format, we can't do.

Reply:
Ok, then it looks like the best what we can do is remove all the .csv and deleted images by their names. Then reexport the whole dataset.

Reply:
This I can do, but wouldn't it be safer to only delete the old annotation files _after_ we're done writing the new ones (which is what I'm doing in the current version)? That way there won't be data loss if, for example, an unexpected exception occurs after the old files are deleted.

> and deleted images by their names

There's actually an unexpected problem with that. If an item is deleted, then we don't know what extension its image had, so we can't actually delete it. As far as I can see, this problem affects your implementations, as well.

Reply:
> This I can do, but wouldn't it be safer to only delete the old annotation files after we're done writing the new ones (which is what I'm doing in the current version)? That way there won't be data loss if, for example, an unexpected exception occurs after the old files are deleted.

The only way of protection from such problems I can come up with is writing to new files, then renaming to the old names. Because once you opened a file with `open(..., 'w')`, the old data is lost.

> we don't know what extension its image had, so we can't actually delete it. As far as I can see, this problem affects your implementations, as well.

Yes, I know about this. A way to fix this could be using of image file search by name, and it was implemented in some format, but it looks like overkill right now. Maybe, it could be a next update for the patching functionality.

Reply:
> The only way of protection from such problems I can come up with is writing to new files, then renaming to the old names.

Okay. I could implement this quite easily (since I had already centralized the writing of annotations), so I did.

> Yes, I know about this. A way to fix this could be using of image file search by name, and it was implemented in some format, but it looks like overkill right now. Maybe, it could be a next update for the patching functionality.

Okay. I implemented your solution for deleting images for now.